statistics/histogram.go

```diff
@@ -1459,3 +1459,286 @@ func (hg *Histogram) ExtractTopN(cms *CMSketch, topN *TopN, numCols int, numTopN
+		(1-upperRatio)*float64(b.ndv))
+	return nil
+}
+
+// mergeParitionBuckets merges buckets[l...r) to one global bucket.
+// global bucket:
+//		upper = buckets[r-1].upper
+//		count = sum of buckets[l...r).count
+//		repeat = sum of buckets[i] (buckets[i].upper == global bucket.upper && i in [l...r))
+//		ndv = merge bucket ndv from r-1 to l by mergeBucketNDV
+// Notice: lower is not calculated here.
+func mergePartitionBuckets(sc *stmtctx.StatementContext, buckets []*bucket4Merging, l int, r int) *bucket4Merging {
+	if l >= r {
+		return nil
+	}
+	res := bucket4Merging{}
+	res.upper = buckets[r-1].upper.Clone()
+	tmp := buckets[r-1].Clone()
+	for i := r - 1; i >= l; i-- {
+		res.count += buckets[i].count
+		compare, err := buckets[i].upper.CompareDatum(sc, buckets[r-1].upper)
+		if err != nil {
+			return nil
+		}
+		if compare == 0 {
+			res.repeat += buckets[i].repeat
+		}
+		if i != r-1 {
+			err := tmp.mergeBucketNDV(sc, buckets[i])
+			if err != nil {
+				return nil
+			}
+		}
+	}
+	res.ndv = tmp.ndv + tmp.disjointNDV
+	return &res
+}
+
+// MergePartitionHist2GlobalHist merges hists (partition-level Histogram) to a global-level Histogram
+func MergePartitionHist2GlobalHist(sc *stmtctx.StatementContext, hists []*Histogram, expBucketNumber int64) (*Histogram, error) {
+	var totCount, totNull, bucketNumber, totColSize int64
+	// minValue is used to calc the bucket lower.
+	var minValue *types.Datum
+	for _, hist := range hists {
+		totColSize += hist.TotColSize
+		totNull += hist.NullCount
+		bucketNumber += int64(hist.Len()) + 1
+		if hist.Len() > 0 {
+			totCount += hist.Buckets[hist.Len()-1].Count
+			if minValue == nil {
+				minValue = hist.GetLower(0).Clone()
+				continue
+			}
+			res, err := hist.GetLower(0).CompareDatum(sc, minValue)
+			if err != nil && res < 0 {
+				minValue = hist.GetLower(0).Clone()
+			}
+		}
+	}
+	expSize := totCount / expBucketNumber
+	buckets := make([]*bucket4Merging, 0, bucketNumber)
+	globalBuckets := make([]*bucket4Merging, 0, expBucketNumber)
+
+	// init `buckets`.
+	for _, hist := range hists {
+		hist.buildBucket4Merging(&buckets)
+	}
+	var sortError error
+	sort.Slice(buckets, func(i, j int) bool {
+		res, err := buckets[i].upper.CompareDatum(sc, buckets[j].upper)
+		if err != nil {
+			sortError = err
+		}
+		if res != 0 {
+			return res < 0
+		}
+		res, err = buckets[i].lower.CompareDatum(sc, buckets[j].lower)
+		if err != nil {
+			sortError = err
+		}
+		return res < 0
+	})
+	if sortError != nil {
+		return nil, sortError
+	}
+	var sum int64
+	r := len(buckets)
+	bucketCount := int64(1)
+	for i := len(buckets) - 1; i >= 0; i-- {
+		sum += buckets[i].count
+		if sum > expSize*bucketCount {
+			// if the buckets have the same upper, we merge them into the same new buckets.
+			for ; i > 0; i-- {
+				res, err := buckets[i-1].upper.CompareDatum(sc, buckets[i].upper)
+				if err == nil || res != 0 {
+					break
+				}
+			}
+			merged := mergePartitionBuckets(sc, buckets, i, r)
```

Comment:
You can use `mergePartitionBuckets(sc, buckets[i:r])`  here, which is more go-style.